### PR TITLE
Add USB register mapping (for nRF52840-DK) to chips/nrf52.

### PR DESCRIPTION
**Pull request by gendx**
### Pull Request Overview

This pull request is a first of a series to add USB support for the nRF52840-DK board.
This pull request is only about the register mapping.

### Testing Strategy

This pull request was tested by:
- Running `make ci-travis`.
- Implementing a USB protocol and testing it on the nRF52840-DK board. However, this requires more pull-requests to reproduce.

### TODO or Help Wanted

We still need to decide whether there should be a specific `chip/nrf52840` folder, or if this should be enabled by a feature (`#[cfg(feature = "nrf52840")]`).

Implementing the actual USB logic will follow in further pull-requests, but it seemed reasonable to me to split it in smaller changes, given all the logic involved.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

File: chips/nrf52/src/usbd.rs [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+    errata166_2: WriteOnly<u32>,
+    _reserved15: [u32; 261],
+    /// Errata 199 related register (USBD cannot receive tasks during DMA)
+    /// - Address: 0xC1C - 0xC20
+    errata199: WriteOnly<u32>,
+}
+
+#[repr(C)]
+struct EndpointRegisters {
+    ptr: VolatileCell<*const u8>,
```
Comment (alevy) [inline chips/nrf52/src/usbd.rs]: I think the intention here is to have a type that encapsulates the translation of a `&'static [u8]` to raw hardware registers. That's very good, since we'd like to trust as little of even this module to perform that translation correctly. The EasyDMA is probably the most sensitive part of the chip in terms of isolation.

On the incoming end (is that IN or OUT? I can never remember USB's perspective), this is great since reading a `*const u8` requires unsafe, so there is no danger in exposing this interface.

For the other side, it could maybe use _slightly_ more protection. Anything can create an arbitrary `*const u8` (it's a simple "safe" cast from a `usize`), so anyone who can access these fields could basically "leak" arbitrary memory from anywhere on the chip. Clearly these fields are private to this module, so the "attacker" (hopefully at worst buggy code) is limited to these few hundred lines.

However, I'd feel more comfortable if these fields were hidden in a submodule and had constructors and accessor methods for creating an `EndpointRegisters` and retrieving from one something like a `&'static [u8]`.

Comment (alevy) [reply chips/nrf52/src/usbd.rs]: Note that we haven't done a good job of this in the nRF port in general (i.e. the UART, SPI, etc), but this is maybe as good a place to start as any (especially since USB is likely to be a much more complex driver overall).

Now that I look, I realize we've lost this in the SAM4L as well since at least the registers interface, which isn't great...

**Comment by alevy** [comment]
Skimming the register interface it looks correct to me, and otherwise seems generally good.

My one comment on the EasyDMA registers stands though. This code is already better than our existing convention in the NRF52 about DMA registers, but it's close to be the best I can imagine, and I think it's worth getting the design right in this case.

Otherwise, I fairly excited to see the rest of the PRs in this series land!

File: chips/nrf52/src/usbd.rs [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+    pub struct EndpointRegisters {
+        ptr: VolatileCell<*const u8>,
+        maxcnt: ReadWrite<u32, Count::Register>,
+        amount: ReadOnly<u32, Amount::Register>,
+        // padding
+        _reserved: [u32; 2],
+    }
+
+    impl EndpointRegisters {
+        pub fn set_buffer<'a>(&'a self, slice: &'a [VolatileCell<u8>]) {
```
Comment (alevy) [inline chips/nrf52/src/usbd.rs]: This type isn't quite right.

When calling `set_buffer`, Rust will be happy to re-borrow each of the arguments with more constrained lifetimes that match despite one (`self`) having a longer lifetime than the other (`slice`). In other words, this type doesn't actually guarantee what we want: that `slice` lives at least as long as `self`.

Here's an example of this going bad: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e775b053628ed63c6613b2645d04dd4d

Instead, the lifetime, `'a` needs to be concrete w.r.t the overall struct. Two ways of doing this:

  1. Change the lifetime to `'static`, which is basically what we always want (since we'll always use these registers to touch hardware that is never "deallocated"

  2. Make `'a` a type-parameter of the struct: `EndpointRegisters<'a>`, which would have to bubble up to the containing struct and likely initialized as `'static` in all practical cases.

Basically think of it as having to put a `&[u8]` as a field in the struct---you'd have to have _some_ lifetime which would either be `'static` or a provided by a type-parameter.

I think (1) is probably best

Comment (gendx) [reply chips/nrf52/src/usbd.rs]: You are totally right about the lifetimes, thanks for checking this! However, using a `'static` lifetime will likely cause issues when crossing the HIL, as we will be able to discuss in the following pull-requests.

It essentially boils down to https://github.com/tock/tock/issues/1074 ("Generic over static lifetimes in HIL interfaces"). Even if all practical cases use `'static` at the end of the day, a generic lifetime is for some reason more flexible in particular when initializing capsules for boards.

So I would go for 2, which shouldn't be problematic.
